lc0/src/neural/network_cudnn.cu

```diff
@@ -0,0 +1,836 @@
+    }
+    else
+    {
+        reportCUDNNErrors(cudnnConvolutionBiasActivationForward(g_cudnn, &alpha, inTensorDesc, input, filterDesc, weights, convDesc,
+            convAlgo, scratch, CUDA_SCRATCH_SIZE, &beta, outTensorDesc, output, biasDesc, biases,
+            activation, outTensorDesc, output));
+    }
+}
+
+ConvLayer::~ConvLayer()
+{
+    cudaFree(weights);
+    cudaFree(biases);
+}
+
+BNLayer::BNLayer(BaseLayer *ip, bool relu) :
+    BaseLayer(ip->getC(), ip->getH(), ip->getW(), ip),
+    useRelu(relu)
+{
+    size_t weightSize = bpe * C;
+
+    cudaMalloc(&means, weightSize);
+    cudaMalloc(&variances, weightSize);
+}
+
+void BNLayer::loadWeights(void *cpuMeans, void *cpuVar)
+{
+    size_t weightSize = bpe * C;
+    cudaMemcpyAsync(means, cpuMeans, weightSize, cudaMemcpyHostToDevice);
+    cudaMemcpyAsync(variances, cpuVar, weightSize, cudaMemcpyHostToDevice);
+}
+
+void BNLayer::eval(void *output, void *input, void *input2, void *scratch)
+{
+    batchNormForward((float *)output, (float *)input, (float *)input2, N, C, H, W, (float *)means, (float *)variances, useRelu);
+}
+
+BNLayer::~BNLayer()
+{
+    cudaFree(means);
+    cudaFree(variances);
+}
+
+FCLayer::FCLayer(BaseLayer *ip, int C, int H, int W, bool relu, bool bias, bool tanh) :
+    BaseLayer(C, H, W, ip),
+    useRelu(relu),
+    useBias(bias),
+    useTanh(tanh)
+{
+    size_t weightSize = bpe * C * H * W * ip->getC() * ip->getH() * ip->getW();
+    size_t biasSize = bpe * C * H * W;
+    cudaMalloc(&weights, weightSize);
+    if (useBias)
+    {
+        cudaMalloc(&biases, biasSize);
+    }
+    else
+    {
+        biases = NULL;
+    }
+}
+
+void FCLayer::loadWeights(void *cpuWeight, void *cpuBias)
+{
+    size_t weightSize = bpe * C * H * W * input->getC() * input->getH() * input->getW();
+    cudaMemcpyAsync(weights, cpuWeight, weightSize, cudaMemcpyHostToDevice);
+    if (useBias)
+    {
+        size_t biasSize = bpe * C * H * W;
+        cudaMemcpyAsync(biases, cpuBias, biasSize, cudaMemcpyHostToDevice);
+    }
+}
+
+void FCLayer::eval(void *outputTensor, void *inputTensor, void *input2, void *scratch)
+{
+    float alpha = 1.0f, beta = 0.0f;
+    int numOutputs = C * H * W;
+    int numInputs = input->getC() * input->getH() * input->getW();
+
+    if (fp16)
+    {
+        // TODO: implement this!
+        assert(0);
+    }
+    else
+    {
+        cublasSgemm(g_cublas, CUBLAS_OP_T, CUBLAS_OP_N,
+            numOutputs, N, numInputs,
+            &alpha,
+            (float*)weights, numInputs,
+            (float*)inputTensor, numInputs,
+            &beta,
+            (float*)outputTensor, numOutputs);
+
+        if (useBias || useRelu || useTanh)
+        {
+            addVectors((float*)outputTensor, (float*)biases, (float*)outputTensor, numOutputs * N, numOutputs, numOutputs * N, useRelu, useTanh);
+        }
+    }
+
+}
+
+FCLayer::~FCLayer()
+{
+    cudaFree(weights);
+    cudaFree(biases);
+}
+
+
+namespace lczero {
```

Comment:
Would it be possible to put all the locally defined classes (above) into the anonymous namespace?
I'm not sure if that's compatible with CUDA compiler, but if it is, that's surely something to do, in order not to clutter global namespace.

Also if that's possible, consider moving this line to the very top of the file.

Reply:
moved everything inside lczero namespace